nixos/modules/installer/tools/nixos-generate-config/templates/hardware-configuration.nix.tera

```diff
@@ -0,0 +1,29 @@
+# Do not modify this file!  It was generated by ‘nixos-generate-config’
+# and may be overwritten by future invocations.  Please make changes
+# to /etc/nixos/configuration.nix instead.
+{ config, lib, pkgs, modulesPath, ... }:
+
+{
+  imports = {{ imports }};
+
+  boot.initrd.availableKernelModules = {{ initrd_available_kernel_modules }};
+  boot.initrd.kernelModules = {{ initrd_kernel_modules }};
+  boot.kernelModules = {{ kernel_modules }};
+  boot.extraModulePackages = {{ module_packages }};
+{%- for fs in filesystems %}
```

Comment:
I would say this is arguable readable than the concatenation that was used before.

Reply:
While I don't wish to oppose the use of this Tera, I wonder whether the standard `fmt` module would be enough.  Was it already tried and determined not to suffice?  If not, here's [a rough but runnable sketch in the Rust Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=4381228a28c336c1e8bce0829d47268b) of how this could look.

Reply:
I don't think it support loops? This is actually my only concern because we need to do string concatenation again which is not very readable.

Reply:
I had hoped it would be comparably readable to Tera, but it didn't turn out as well as I'd hoped.  The explicit string concatenation (line 56) could be replaced with [`Itertools::format`](https://docs.rs/itertools/0.9.0/itertools/trait.Itertools.html#method.format) ([adjusted Playground sketch](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=1249791d41d90a737089f270813f2d25)).  Using that itertools method for the loop (and also to simplify the `Display` implementation for `NixList`) increases the number of lines of assembly code to which this minimal example compiles by 11%; presumably the effect would be proportionally smaller for a real program — but then I don't wager Tera will be large enough to be a problem.

Perhaps the largest difference to me between the two approaches (to be clear, I don't mean to say either is superior for it) is that in the Tera templates one has the structure of the output file and loop logic inlined together whereas with `fmt` parts of the output file such as the list of filesystems necessarily get broken out, like smaller functions being factored out of a larger function.

Reply:
I think the size increase of tera is also due to other crates being linked in, which might end up using later as well i.e. serde. It's good to have an alternative but first I would prioritize finishing the prototype first.

Reply:
Is there a checklist tracking the progress of this prototype?

I've looked at the original post at the top of the PR but it's unclear what is and isn't complete (or at least implemented and might need future tweaks)

Reply:
It's not usable yet. I will post updates as I add new features.